File: kmp/monero-net/src/commonTest/kotlin/io/monero/net/WalletRpcClientTest.kt
package io.monero.net

import io.ktor.client.*
import io.ktor.client.engine.mock.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.util.*
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlin.test.Test
import io.kotest.matchers.shouldBe

class WalletRpcClientTest {

    private fun makeClient(jsonResponse: String): WalletRpcClient {
        val engine = MockEngine { request ->
            request.url.encodedPath shouldBe "/json_rpc"
            respond(
                content = jsonResponse,
                status = HttpStatusCode.OK,
                headers = headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString())
            )
        }

        val json = Json { ignoreUnknownKeys = true; isLenient = true }

        val httpClient = HttpClient(engine) {
            install(ContentNegotiation) {
                json(json)
            }
        }

        return HttpWalletRpcClient(WalletRpcConfig(host = "localhost", port = 18082), httpClient)
    }

    @Test
    fun `getTxKey parses tx_key`() = runTest {
        val client = makeClient("""{"jsonrpc":"2.0","id":"0","result":{"tx_key":"abc"}}""")
        client.getTxKey("txid") shouldBe "abc"
        client.close()
    }

    @Test
    fun `checkTxProof parses result`() = runTest {
        val client = makeClient(
            """{"jsonrpc":"2.0","id":"0","result":{"confirmations":482,"good":true,"in_pool":false,"received":1000}}"""
        )
        val result = client.checkTxProof(
            txid = "txid",
            address = "addr",
            signature = "sig",
            message = "msg"
        )
        result.good shouldBe true
        result.confirmations shouldBe 482
        result.inPool shouldBe false
        result.received shouldBe 1000
        client.close()
    }

    @Test
    fun `checkReserveProof parses result`() = runTest {
        val client = makeClient(
            """{"jsonrpc":"2.0","id":"0","result":{"good":true,"spent":0,"total":123}}"""
        )
        val result = client.checkReserveProof(address = "addr", signature = "sig")
        result.good shouldBe true
        result.total shouldBe 123
        client.close()
    }

    @Test
    fun `sign and verify parse results`() = runTest {
        val signClient = makeClient("""{"jsonrpc":"2.0","id":"0","result":{"signature":"SigV1xxx"}}""")
        signClient.sign("hello") shouldBe "SigV1xxx"
        signClient.close()

        val verifyClient = makeClient("""{"jsonrpc":"2.0","id":"0","result":{"good":true}}""")
        verifyClient.verify(data = "hello", address = "addr", signature = "SigV1xxx") shouldBe true
        verifyClient.close()
    }

    @Test
    fun `multisig responses parse`() = runTest {
        val isMultisigClient = makeClient(
            """{"jsonrpc":"2.0","id":"0","result":{"multisig":true,"ready":true,"threshold":2,"total":3}}"""
        )
        val status = isMultisigClient.isMultisig()
        status.multisig shouldBe true
        status.threshold shouldBe 2
        isMultisigClient.close()

        val prepareClient = makeClient(
            """{"jsonrpc":"2.0","id":"0","result":{"multisig_info":"MultisigV1..."}}"""
        )
        prepareClient.prepareMultisig().multisigInfo shouldBe "MultisigV1..."
        prepareClient.close()

        val signMultisigClient = makeClient(
            """{"jsonrpc":"2.0","id":"0","result":{"tx_data_hex":"deadbeef","tx_hash_list":["h1","h2"]}}"""
        )
        val signed = signMultisigClient.signMultisig("deadbeef")
        signed.txDataHex shouldBe "deadbeef"
        signed.txHashList.size shouldBe 2
        signMultisigClient.close()
    }
}

File: dart/test/network/wallet_rpc_client_test.dart
import 'package:monero_dart/src/network/wallet_rpc_client.dart';
import 'package:test/test.dart';

void main() {
  group('WalletRpcClient', () {
    test('can be instantiated with defaults', () {
      final client = WalletRpcClient();
      expect(client.host, equals('localhost'));
      expect(client.port, equals(18082));
      expect(client.useSsl, isFalse);
      client.close();
    });

    test('can be customized', () {
      final client = WalletRpcClient(
        host: 'wallet.example.com',
        port: 18083,
        useSsl: true,
        username: 'user',
        password: 'pass',
        timeout: Duration(seconds: 60),
      );
      expect(client.host, equals('wallet.example.com'));
      expect(client.port, equals(18083));
      expect(client.useSsl, isTrue);
      client.close();
    });
  });

  group('Wallet RPC DTOs', () {
    test('CheckTxProofResult.fromJson parses correctly', () {
      final json = {
        'confirmations': 482,
        'good': true,
        'in_pool': false,
        'received': 1000000000000,
      };
      final result = CheckTxProofResult.fromJson(json);
      expect(result.good, isTrue);
      expect(result.confirmations, equals(482));
      expect(result.inPool, isFalse);
    });

    test('CheckReserveProofResult.fromJson parses BigInt fields', () {
      final json = {
        'good': true,
        'spent': 0,
        'total': 100000000000,
      };
      final result = CheckReserveProofResult.fromJson(json);
      expect(result.good, isTrue);
      expect(result.total.toString(), equals('100000000000'));
    });

    test('IsMultisigResult.fromJson parses correctly', () {
      final json = {
        'multisig': true,
        'ready': true,
        'threshold': 2,
        'total': 3,
      };
      final result = IsMultisigResult.fromJson(json);
      expect(result.multisig, isTrue);
      expect(result.threshold, equals(2));
    });

    test('SignMultisigResult.fromJson parses correctly', () {
      final json = {
        'tx_data_hex': 'deadbeef',
        'tx_hash_list': ['h1', 'h2'],
      };
      final result = SignMultisigResult.fromJson(json);
      expect(result.txDataHex, equals('deadbeef'));
      expect(result.txHashList.length, equals(2));
    });
  });
}
